From 37c0a6113967619af24d58c3730afcf57f3bfbad Mon Sep 17 00:00:00 2001
From: Rabeeh Khoury <rabeeh@solid-run.com>
Date: Sun, 17 Jan 2021 17:14:06 +0200
Subject: [PATCH] lx2160acex7: add poweroff and disable fan full speed

1. Add poweroff command; when using Clearfog CX or HoneyComb carrier
boards, the S5# signal is connected to the power button controller that
cuts off the '12v' supply from the COM module.

2. In board_fix_fdt, just before booting the kernel the fan full speed
signal becomes inactive so that the kernel AMC6821 driver would take
over and enable the auto thermal and pwm adjustments.

Signed-off-by: Rabeeh Khoury <rabeeh@solid-run.com>
---
 arch/arm/include/asm/arch-fsl-layerscape/immap_lsch3.h | 3 +++
 board/solidrun/lx2160a/lx2160a.c                       | 8 +++++++-
 configs/lx2160acex7_tfa_defconfig                      | 2 +-
 3 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/arch/arm/include/asm/arch-fsl-layerscape/immap_lsch3.h b/arch/arm/include/asm/arch-fsl-layerscape/immap_lsch3.h
index dc740c999d..ab6297dae2 100644
--- a/arch/arm/include/asm/arch-fsl-layerscape/immap_lsch3.h
+++ b/arch/arm/include/asm/arch-fsl-layerscape/immap_lsch3.h
@@ -65,6 +65,9 @@
 #define I2C7_BASE_ADDR				(CONFIG_SYS_IMMR + 0x01060000)
 #define I2C8_BASE_ADDR				(CONFIG_SYS_IMMR + 0x01070000)
 #endif
+#define GPIO3_BASE_ADDR				(CONFIG_SYS_IMMR + 0x01320000)
+#define GPIO3_GPDIR_ADDR			(GPIO3_BASE_ADDR + 0x0)
+#define GPIO3_GPDAT_ADDR			(GPIO3_BASE_ADDR + 0x8)
 #define GPIO4_BASE_ADDR				(CONFIG_SYS_IMMR + 0x01330000)
 #define GPIO4_GPDIR_ADDR			(GPIO4_BASE_ADDR + 0x0)
 #define GPIO4_GPDAT_ADDR			(GPIO4_BASE_ADDR + 0x8)
diff --git a/board/solidrun/lx2160a/lx2160a.c b/board/solidrun/lx2160a/lx2160a.c
index 975431fd53..0148ce2a48 100644
--- a/board/solidrun/lx2160a/lx2160a.c
+++ b/board/solidrun/lx2160a/lx2160a.c
@@ -358,7 +358,13 @@ int ft_board_setup(void *blob, bd_t *bd)
 	fdt_fsl_mc_fixup_iommu_map_entry(blob);
 	fdt_fixup_board_enet(blob);
 #endif
-
+	printf ("Releasing fan controller full speed gpio\n");
+	/*
+	 * Set the GPIO to be input; the on-COM pullup will disable the full speed
+	 * signal.
+	 */
+	out_le32(GPIO3_GPDIR_ADDR, (~(1 << 29) &
+		in_le32(GPIO3_GPDIR_ADDR)));
 	return 0;
 }
 #endif
diff --git a/configs/lx2160acex7_tfa_defconfig b/configs/lx2160acex7_tfa_defconfig
index 8e96265843..837d6070dd 100644
--- a/configs/lx2160acex7_tfa_defconfig
+++ b/configs/lx2160acex7_tfa_defconfig
@@ -68,7 +68,7 @@ CONFIG_PHY_ATHEROS=y
 CONFIG_PCIE_LAYERSCAPE=y
 CONFIG_DM_RTC=y
 CONFIG_RTC_PCF2127=y
-
+CONFIG_CMD_POWEROFF=y
 CONFIG_DM_SCSI=y
 CONFIG_DM_SERIAL=y
 CONFIG_SPI=y
-- 
2.25.1

